File: examples/typescript/virtual_mouse.ts
import { ViiperClient, ViiperDevice, Mouse, Types } from "viiperclient";

const { MouseInput, Btn_ } = Mouse;

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

// Minimal example: ensure a bus, create a mouse device, stream inputs, clean up on exit.
async function main() {
  if (process.argv.length < 3) {
    console.log("Usage: node virtual_mouse.js <api_addr>");
    console.log("Example: node virtual_mouse.js localhost:3242");
    process.exit(1);
  }

  const addr = process.argv[2];
  const [host, portStr] = addr.split(':');
  const port = portStr ? parseInt(portStr, 10) : 3242;
  const client = new ViiperClient(host, port);

  // Find or create a bus
  const busesResp = await client.buslist();
  let busID: number;
  let createdBus = false;
  
  if (busesResp.buses.length === 0) {
    let createErr: any;
    busID = 0;
    for (let tryBus = 1; tryBus <= 100; tryBus++) {
      try {
        const r = await client.buscreate(tryBus);
        busID = r.busId;
        createdBus = true;
        break;
      } catch (err) {
        createErr = err;
      }
    }
    if (busID === 0) {
      console.error(`BusCreate failed: ${createErr}`);
      process.exit(1);
    }
    console.log(`Created bus ${busID}`);
  } else {
    busID = Math.min(...busesResp.buses);
    console.log(`Using existing bus ${busID}`);
  }

  // Add device and connect to stream in one call
  let dev: ViiperDevice;
  let deviceDevId: string;
  try {
    const req: Types.DeviceCreateRequest = { type: "mouse" };
    const { device, response: addResp } = await client.addDeviceAndConnect(busID, req);
    dev = device;
    deviceDevId = addResp.devId;
    console.log(`Created and connected to device ${deviceDevId} on bus ${busID}`);
  } catch (err) {
    console.error(`AddDeviceAndConnect error: ${err}`);
    if (createdBus) {
      await client.busremove(busID).catch(() => {});
    }
    process.exit(1);
  }

  // Cleanup function
  const cleanup = async () => {
    try {
      dev.close();
      await client.busdeviceremove(busID, deviceDevId);
      console.log(`Removed device ${deviceDevId}`);
    } catch (err) {
      console.error(`DeviceRemove error: ${err}`);
    }
    if (createdBus) {
      try {
        await client.busremove(busID);
        console.log(`Removed bus ${busID}`);
      } catch (err) {
        console.error(`BusRemove error: ${err}`);
      }
    }
  };

  // Send a short movement once every 3 seconds for easy local testing.
  // Followed by a short click and a single scroll notch.
  let dir = 1;
  const step = 50; // move diagonally by 50 px in X and Y
  let running = true;

  console.log("Every 3s: move diagonally by 50px (X and Y), then click and scroll. Press Ctrl+C to stop.");

  const interval = setInterval(async () => {
    if (!running) return;
    
    try {
      // Move diagonally: (+step,+step) then (-step,-step) next tick
      const dx = step * dir;
      const dy = step * dir;
      dir *= -1;

      // One-shot movement report (diagonal)
      const move = new MouseInput({ Buttons: 0, Dx: dx, Dy: dy, Wheel: 0, Pan: 0 });
      await dev.send(move);
      console.log(`→ Moved mouse dx=${dx} dy=${dy}`);

      // Zero state shortly after to keep movement one-shot (harmless safety)
      await sleep(30);
      const zero = new MouseInput({ Buttons: 0, Dx: 0, Dy: 0, Wheel: 0, Pan: 0 });
      await dev.send(zero);

      // Simulate a short left click: press then release
      await sleep(50);
      const press = new MouseInput({ Buttons: Btn_.Left, Dx: 0, Dy: 0, Wheel: 0, Pan: 0 });
      await dev.send(press);
      await sleep(60);
      const rel = new MouseInput({ Buttons: 0x00, Dx: 0, Dy: 0, Wheel: 0, Pan: 0 });
      await dev.send(rel);
      console.log("→ Clicked (left)");

      // Simulate a short scroll: one notch upwards
      await sleep(50);
      const scr = new MouseInput({ Buttons: 0, Dx: 0, Dy: 0, Wheel: 1, Pan: 0 });
      await dev.send(scr);
      await sleep(30);
      const scr0 = new MouseInput({ Buttons: 0, Dx: 0, Dy: 0, Wheel: 0, Pan: 0 });
      await dev.send(scr0);
      console.log("→ Scrolled (wheel=+1)");
    } catch (err) {
      console.error(`Write error: ${err}`);
      running = false;
      clearInterval(interval);
      await cleanup();
      process.exit(1);
    }
  }, 3000);

  dev.on("error", async (err: Error) => {
    console.error(`Stream error: ${err}`);
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(1);
  });

  dev.on("end", async () => {
    console.log("Stream ended by server");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });

  // Handle signals for graceful shutdown
  process.on("SIGINT", async () => {
    console.log("Signal received, stopping…");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });
  process.on("SIGTERM", async () => {
    console.log("Signal received, stopping…");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/cpp/virtual_mouse.cpp
#define VIIPER_JSON_INCLUDE <nlohmann/json.hpp>
#define VIIPER_JSON_NAMESPACE nlohmann
#define VIIPER_JSON_TYPE json

#include <viiper/viiper.hpp>
#include <iostream>
#include <thread>
#include <chrono>
#include <csignal>
#include <atomic>

std::atomic<bool> running{true};

void signal_handler(int) {
    running = false;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <api_addr>\n";
        std::cerr << "Example: " << argv[0] << " localhost:3242\n";
        return 1;
    }

    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    const std::string addr = argv[1];
    const auto colon_pos = addr.find(':');
    const std::string host = addr.substr(0, colon_pos);
    const std::uint16_t port = colon_pos != std::string::npos
        ? static_cast<std::uint16_t>(std::stoul(addr.substr(colon_pos + 1)))
        : 3242;

    viiper::ViiperClient client(host, port);

    // Find or create a bus
    std::uint32_t bus_id;
    bool created_bus = false;

    auto buses_result = client.buslist();
    if (buses_result.is_error()) {
        std::cerr << "BusList error: " << buses_result.error().to_string() << "\n";
        return 1;
    }

    if (buses_result.value().buses.empty()) {
        auto create_result = client.buscreate(std::nullopt);
        if (create_result.is_error()) {
            std::cerr << "BusCreate failed: " << create_result.error().to_string() << "\n";
            return 1;
        }
        bus_id = create_result.value().busid;
        created_bus = true;
        std::cout << "Created bus " << bus_id << "\n";
    } else {
        bus_id = buses_result.value().buses[0];
        std::cout << "Using existing bus " << bus_id << "\n";
    }

    // Add device
    auto device_result = client.busdeviceadd(bus_id, {.type = "mouse"});
    if (device_result.is_error()) {
        std::cerr << "AddDevice error: " << device_result.error().to_string() << "\n";
        if (created_bus) {
            client.busremove(bus_id);
        }
        return 1;
    }
    auto device_info = std::move(device_result.value());

    // Connect to device stream
    auto stream_result = client.connectDevice(device_info.busid, device_info.devid);
    if (stream_result.is_error()) {
        std::cerr << "ConnectDevice error: " << stream_result.error().to_string() << "\n";
        client.busdeviceremove(device_info.busid, device_info.devid);
        if (created_bus) {
            client.busremove(bus_id);
        }
        return 1;
    }
    auto stream = std::move(stream_result.value());

    std::cout << "Created and connected to device " << device_info.devid
              << " on bus " << device_info.busid << "\n";

    std::cout << "Every 3s: move diagonally by 50px (X and Y), then click and scroll. Press Ctrl+C to stop.\n";

    // Send a short movement once every 3 seconds for easy local testing.
    // Followed by a short click and a single scroll notch.
    int dir = 1;
    constexpr int step = 50; // move diagonally by 50 px in X and Y

    while (running && stream->is_connected()) {
        // Move diagonally: (+step,+step) then (-step,-step) next tick
        std::int16_t dx = static_cast<std::int16_t>(step * dir);
        std::int16_t dy = static_cast<std::int16_t>(step * dir);
        dir *= -1;

        // One-shot movement report (diagonal)
        auto send_result = stream->send(viiper::mouse::Input{
            .buttons = 0,
            .dx = dx,
            .dy = dy,
            .wheel = 0,
            .pan = 0,
        });
        if (send_result.is_error()) {
            std::cerr << "Write error: " << send_result.error().to_string() << "\n";
            break;
        }
        std::cout << "→ Moved mouse dx=" << static_cast<int>(dx) << " dy=" << static_cast<int>(dy) << "\n";

        // Zero state shortly after to keep movement one-shot (harmless safety)
        std::this_thread::sleep_for(std::chrono::milliseconds(30));
        stream->send(viiper::mouse::Input{.buttons = 0, .dx = 0, .dy = 0, .wheel = 0, .pan = 0});

        // Simulate a short left click: press then release
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        stream->send(viiper::mouse::Input{
            .buttons = viiper::mouse::Btn_Left,
            .dx = 0, .dy = 0, .wheel = 0, .pan = 0,
        });
        std::this_thread::sleep_for(std::chrono::milliseconds(60));
        stream->send(viiper::mouse::Input{.buttons = 0, .dx = 0, .dy = 0, .wheel = 0, .pan = 0});
        std::cout << "→ Clicked (left)\n";

        // Simulate a short scroll: one notch upwards
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        stream->send(viiper::mouse::Input{.buttons = 0, .dx = 0, .dy = 0, .wheel = 1, .pan = 0});
        std::this_thread::sleep_for(std::chrono::milliseconds(30));
        stream->send(viiper::mouse::Input{.buttons = 0, .dx = 0, .dy = 0, .wheel = 0, .pan = 0});
        std::cout << "→ Scrolled (wheel=+1)\n";

        std::this_thread::sleep_for(std::chrono::seconds(3));
    }

    // Cleanup
    stream->stop();
    client.busdeviceremove(device_info.busid, device_info.devid);
    if (created_bus) {
        client.busremove(bus_id);
    }

    return 0;
}
